db/migrations/20221114170158_revise_sticky_triggers.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+class ReviseStickyTriggers extends DatabaseMigration {
+  public function migrate() {
+    $sth = db_query("select iid from f_indexes order by iid");
+    echo "DO NOT INTERRUPT, this could take quite some time!\n";
+    while ($i = $sth->fetch() ) {
```

Comment:
this seems super slow.

Can't we change the way this works entirely?

that is to say
1) remove dupes from the sticky tables
2) change the schema to make mid/tid unique

would the existing trigger still be ok?

Reply:
1. Sure that's probably easier. We can do this:
```
$sqlCleanStickyTable = "DELETE t1 FROM . " .$triggertbl . "t1 " .
  "INNER JOIN " .$triggertbl . "t2 " .
  "WHERE t1.sid < t2.sid AND t1.tid = t2.tid;"
echo "Cleaning table $tbl\n";
db_exec($sqlCleanStickyTable);
```

2. Updating the schema for sticky would work to add a unique constraint, but the trigger would need to be updated to avoid duplicate exceptions. I would want to remove the mid column if we were updating the f_sticky schema. The trigger should get updated to avoid any opportunities of duplicate key exceptions. 
```
  $sqlTrigger = "drop trigger if exists trigger_" . $tbl . "_sticky_update; " .
    "drop trigger if exists trigger_f" . $i['iid'] . "_sticky_update; " .
    "create trigger trigger_f" . $i['iid'] . "_sticky_update " .
    "after update on " . $triggertbl . " " .
    "for each row " .
    "begin " .
    "if new.flags like '%STICKY%' then " .
    "if new.flags <> old.flags then " .
    "insert into " . $tbl . "(tid) values (new.tid) " .
    "on duplicate key update tid=tid;" .
    "end if; " .
    "else " .
    "delete from " . $tbl . " where tid = new.tid; " .
    "end if; " .
    "end ";
```
But if were adding the unique constraint the trigger wouldn't need to confirm old flag value since we can just update the row setting tid = tid and could be simplified to:
```
  $sqlTrigger = "drop trigger if exists trigger_" . $tbl . "_sticky_update; " .
    "drop trigger if exists trigger_f" . $i['iid'] . "_sticky_update; " .
    "create trigger trigger_f" . $i['iid'] . "_sticky_update " .
    "after update on " . $triggertbl . " " .
    "for each row " .
    "begin " .
    "if new.flags like '%STICKY%' then " .
    "insert into " . $tbl . "(tid) values (new.tid) " .
    "on duplicate key update tid=tid;" .
    "else " .
    "delete from " . $tbl . " where tid = new.tid; " .
    "end if; " .
    "end ";
```

Reply:
i was thinking more literally, just remove the duplicate entries from the f_sticky tables directly rather than rebuilding them (let alone have the trigger do it).

Reply:
Sorry what i couldn't figure out is the relationship between `$triggertbl` and `f_stickyX` - is it *literally* the name of the sticky table?

Reply:
Yes, that's the relation. I think a cleaner version would look like the following. It's a bit sloppy on the $triggername since I had a typo on the original forum creation script compared to the migration. 
```
class ReviseStickyTriggers extends DatabaseMigration {
  public function migrate() {
    $sth = db_query("select iid from f_indexes order by iid");
    $forumindexids = $sth->fetchAll();
    $sth->closeCursor();
    echo "DO NOT INTERRUPT, this update should complete shortly\n";
    foreach($forumindexids as $forumindexid) {
      $triggertbl = "f_threads" . $forumindexid['iid'];
      $stickytbl = "f_sticky" . $forumindexid['iid'];
      $tiggername = "trigger_f" . $forumindexid['iid'];

      echo "Updating $tiggername for $triggertbl / $stickytbl\n";
      $sqlTriggerCreation = 
        "drop trigger if exists trigger_" . $stickytbl . "_sticky_update; " . //cleans up old trigger created by migration
        "drop trigger if exists " . $tiggername . "_sticky_update; " . //cleans up old trigger created by new forum
        "create trigger " . $tiggername . "_sticky_update " .
        "after update on " . $triggertbl . " " .
        "for each row " .
        "begin " .
        "if new.flags like '%STICKY%' then " .
        "if new.flags <> old.flags then " .
        "insert into " . $stickytbl . "(tid, mid) values (new.tid, new.mid) " .
        "on duplicate key update tid=tid; " .
        "end if; " .
        "else " .
        "delete from " . $stickytbl . " where tid = new.tid; " .
        "end if; " .
        "end ";
      db_exec($sqlTriggerCreation);

      echo "Cleaning douplicates from sticky table $stickytbl\n";
      $sqlCleanStickyTable = "DELETE t1 FROM . " .$stickytbl . " t1 " .
        "INNER JOIN " .$stickytbl . " t2 " .
        "WHERE t1.sid < t2.sid AND t1.tid = t2.tid;";
      db_exec($sqlCleanStickyTable);
    }
  }
}
```